torch/csrc/ThreadLocalState.h

```diff
@@ -0,0 +1,39 @@
+#pragma once
+
+namespace torch {
+
+// Thread local state that needs to be propagated across different threads
+// during JIT interpreter continuations.
+class ThreadLocalState {
+ public:
+  ThreadLocalState(bool grad_mode_enabled, int64_t dist_autograd_context_id);
+  bool gradModeEnabled() const;
+  int64_t distAutogradContextId() const;
+
+ private:
+  bool grad_mode_enabled_;
+  int64_t dist_autograd_context_id_;
+};
+
+std::shared_ptr<ThreadLocalState> getThreadLocalState();
+
+void setThreadLocalState(const ThreadLocalState& state);
+
+// Guard to set and reset the appropriate thread local state for JIT
+// interpreter continuations.
+class ThreadLocalStateGuard {
+ public:
+  explicit ThreadLocalStateGuard(std::shared_ptr<ThreadLocalState> state) {
+    prev_state_ = getThreadLocalState();
+    setThreadLocalState(*state);
+  }
+
+  ~ThreadLocalStateGuard() {
+    setThreadLocalState(*prev_state_);
+  }
+
+ private:
+  std::shared_ptr<ThreadLocalState> prev_state_;
```

Comment:
Wondering aloud: Do we really need to wrap in a shared ptr, if this is [currently] just an {int64,int32}?

I understand why this makes sense (especially if we add fields in future), but it also feels expensive to add an extra malloc + free, plus several atomic ops, for each jit.fork/wait.

Reply:
Good point, I'll avoid the shared_ptr for now and we can revisit this later when we add more fields.